Enable caching for quarkus:build and declare dependencies as inputs

aloubyansky opened the pull request:
@gsmet here are two commits:
1) enhances the `track-config-changes` impl to also dump dependency checksums in `target/quarkus-<profile-name>-dependency-checksums.txt`, typically the profile name would be `prod`;
2) enables dependency checksums dump for all the IT modules.

fyi @jprinet

aloubyansky committed:
Enable caching of quarkus:build outcome for the main integration test

aloubyansky committed:
Dump all Quarkus app dependencies along with their checksums as part of track-config-changes goal

aloubyansky committed:
Execute quarkus:track-config-changes for all the IT modules to dump quarkus application dependency checksums

gsmet commented on devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java:
```diff
@@ -102,34 +112,60 @@ protected void doExecute() throws MojoExecutionException, MojoFailureException {
+
+            if (dumpDependencies) {
+                final List<String> deps = new ArrayList<>();
+                for (var d : curatedApplication.getApplicationModel().getDependencies(DependencyFlags.DEPLOYMENT_CP)) {
```
For the extension tests, I wonder if we would have to dump the full test deployment classpath as we might add some deployment extensions for testing.

aloubyansky replied:
That's what it does, unless I misunderstood the point.

gsmet committed:
Only compute checksum if snapshot dependency

Non snapshot dependencies are supposed to be immutable so we can skip
computing the checksum.

gsmet committed:
Fix name of config property for consistency

gsmet committed:
Enable config tracking for all ITs

gsmet committed:
Avoid NPE in CodeGenerator.dumpCurrentConfigValues

When we don't have any previous recorded config properties, we should
push an empty Properties. At least, that's what
CodeGenerator.dumpCurrentConfigValues expects.

gsmet committed:
Configure additional input for cached plugins

Add the dumped dependencies as an input for Surefire, Failsafe and the
Quarkus Maven plugin.

gsmet commented on devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java:
```diff
@@ -102,34 +112,64 @@ protected void doExecute() throws MojoExecutionException, MojoFailureException {
-        Properties actualProps;
         final boolean clearPackageTypeSystemProperty = setPackageTypeSystemPropertyIfNativeProfileEnabled();
         try {
             curatedApplication = bootstrapApplication(launchMode);
```
FWIW, I benchmarked things a bit and resolving the dependencies is quite slow. Even for a relatively simple IT as the Hibernate Validator one, it takes ~ 0.8 s.

aloubyansky replied:
It is, this is something i've been looking into for the last few weeks. But there is not much we can do especially for relatively simple projects, since most of the time is spent in the Maven artifact resolver itself.

However, the app model resolution will typically be necessary either way. It's about at which point it's performed. Here, we are moving this step to an earlier phase and it's cached for later phases if they get executed.

gsmet commented on devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java:
```diff
@@ -102,34 +112,64 @@ protected void doExecute() throws MojoExecutionException, MojoFailureException {
+                final List<String> deps = new ArrayList<>();
+                for (var d : curatedApplication.getApplicationModel().getDependencies(DependencyFlags.DEPLOYMENT_CP)) {
+                    StringBuilder entry = new StringBuilder(d.toGACTVString());
+                    if (d.isSnapshot()) {
```
I decided to avoid computing the checksum if the dependency is not a snapshot one. They should be immutable in this case.

aloubyansky replied:
Final versions could still be built locally. For example before releasing but not necessarily then.

gsmet commented on integration-tests/main/pom.xml:
```diff
@@ -23,6 +23,9 @@
         <webjar.jquery-ui.version>1.13.0</webjar.jquery-ui.version>
+
+        <quarkus.config-tracking.enabled>true</quarkus.config-tracking.enabled>
+        <quarkus.native.container-build>true</quarkus.native.container-build>
```
Can you clarify why you added that?

gsmet commented on integration-tests/pom.xml:
```diff
@@ -22,6 +22,9 @@
         <enforce-test-deps-scope.skip>true</enforce-test-deps-scope.skip>
+
+        <quarkus.config-tracking.enabled>true</quarkus.config-tracking.enabled>
+        <quarkus.native.container-build>true</quarkus.native.container-build>
```
Can you clarify why you added `<quarkus.native.container-build>true</quarkus.native.container-build>`?

jprinet replied:
This is required by the [Quarkus caching extension](https://github.com/gradle/develocity-build-config-samples/tree/main/quarkus-build-caching-extension#limitations) to ensure native builds are as reproducible as possible and make caching consistent.

gsmet reviewed (commented):
Things seem to work with the changes I made.

Could you clarify again why it only works the third time? I somehow forgot about it.

gsmet commented:
BTW, it wasn't terribly clear but I added a bunch of commits.

jprinet commented:
> Things seem to work with the changes I made.
> 
> Could you clarify again why it only works the third time? I somehow forgot about it.

This is not true anymore, with the config dump [cached/restored](https://github.com/quarkusio/quarkus/pull/37622/files#diff-d8e36253822cffbdf21e333bcba467345b46e6095927dda98399fe7f21f4679aR1032) on the CI runner, the cache hit would occur on first run if inputs are identical.

You can check the initial explanation [here](https://github.com/quarkusio/quarkus/pull/37258#issuecomment-1827479640)

aloubyansky commented on devtools/maven/src/main/java/io/quarkus/maven/TrackConfigChangesMojo.java:
```diff
@@ -75,7 +75,7 @@ public class TrackConfigChangesMojo extends QuarkusBootstrapMojo {
     /**
      * Dependency dump file
      */
-    @Parameter(property = "quarkus.track-config-changes.dependenciesFile")
```
I suppose to be completely consistent, it would have to be `quarkus.trackConfigChanges.dependenciesFile`

gsmet committed:
Revert "Enable config tracking for all ITs"

This reverts commit b3eeb7c0c049a5c9c75db538446a52b71a45e8e5.

In the end, it is not a very good idea to enable this for all of them.

gsmet committed:
Some refinements to make sure the execution only runs once

zakkak commented on integration-tests/main/pom.xml:
```diff
@@ -23,6 +23,9 @@
         <webjar.jquery-ui.version>1.13.0</webjar.jquery-ui.version>
+
+        <quarkus.config-tracking.enabled>true</quarkus.config-tracking.enabled>
+        <quarkus.native.container-build>true</quarkus.native.container-build>
```
@rsvoboda why is this change necessary?

jprinet replied:
This is required to make the Quarkus build goal cacheable:
https://github.com/gradle/develocity-build-config-samples/tree/main/quarkus-build-caching-extension

zakkak replied:
Good, thanks for the pointer @jprinet
